lib/matplotlib/tests/test_category.py

```diff
@@ -62,196 +67,207 @@ class TestStrCategoryConverter(object):
+        plotter(ax, words, [0, 1])
+        axis_test(ax.xaxis, words)
 
-    def test_plot_update(self):
-        fig, ax = plt.subplots()
-
-        ax.plot(['a', 'b'])
-        ax.plot(['a', 'b', 'd'])
-        ax.plot(['b', 'c', 'd'])
-        fig.canvas.draw()
-
-        labels = ['a', 'b', 'd', 'c']
-        ticks = [0, 1, 2, 3]
-        unit_data = MockUnitData(zip(labels, ticks))
-
-        self.axis_test(ax.yaxis, ticks, labels, unit_data)
-
-    def test_scatter_update(self):
-        fig, ax = plt.subplots()
-
-        ax.scatter(['a', 'b'], [0., 3.])
-        ax.scatter(['a', 'b', 'd'], [1., 2., 3.])
-        ax.scatter(['b', 'c', 'd'], [4., 1., 2.])
-        fig.canvas.draw()
-
-        labels = ['a', 'b', 'd', 'c']
-        ticks = [0, 1, 2, 3]
-        unit_data = MockUnitData(list(zip(labels, ticks)))
-        self.axis_test(ax.xaxis, ticks, labels, unit_data)
+    @pytest.fixture
+    def test_data(self):
+        self.x = ["hello", "happy", "world"]
+        self.xy = [2, 6, 3]
+        self.y = ["Python", "is", "fun"]
+        self.yx = [3, 4, 5]
+
+    @pytest.mark.usefixtures("test_data")
+    @pytest.mark.parametrize("plotter", PLOT_LIST, ids=PLOT_IDS)
+    def test_plot_xaxis(self, ax, test_data, plotter):
+        plotter(ax, self.x, self.xy)
+        axis_test(ax.xaxis, self.x)
+
+    @pytest.mark.usefixtures("test_data")
+    @pytest.mark.parametrize("plotter", PLOT_LIST, ids=PLOT_IDS)
+    def test_plot_yaxis(self, ax, test_data, plotter):
+        plotter(ax, self.yx, self.y)
+        axis_test(ax.yaxis, self.y)
+
+    @pytest.mark.usefixtures("test_data")
+    @pytest.mark.parametrize("plotter", PLOT_LIST, ids=PLOT_IDS)
+    def test_plot_xyaxis(self, ax, test_data, plotter):
+        plotter(ax, self.x, self.y)
+        axis_test(ax.xaxis, self.x)
+        axis_test(ax.yaxis, self.y)
+
+    @pytest.mark.parametrize("plotter", PLOT_LIST, ids=PLOT_IDS)
+    def test_update_plot(self, ax, plotter):
+        plotter(ax, ['a', 'b'], ['e', 'g'])
+        plotter(ax, ['a', 'b', 'd'], ['f', 'a', 'b'])
+        plotter(ax, ['b', 'c', 'd'], ['g', 'e', 'd'])
+        axis_test(ax.xaxis, ['a', 'b', 'd', 'c'])
+        axis_test(ax.yaxis, ['e', 'g', 'f', 'a', 'b', 'd'])
+
+    failing_test_cases = [("mixed", ['A', 3.14]),
+                          ("number integer", ['1', 1]),
+                          ("string integer", ['42', 42]),
+                          ("missing", ['12', np.nan])]
+
+    fids, fvalues = zip(*failing_test_cases)
+
+    PLOT_BROKEN_LIST = [Axes.scatter,
+                        pytest.param(Axes.plot, marks=pytest.mark.xfail),
+                        pytest.param(Axes.bar, marks=pytest.mark.xfail)]
+
+    PLOT_BROKEN_IDS = ["scatter", "plot", "bar"]
+
+    @pytest.mark.parametrize("plotter", PLOT_BROKEN_LIST, ids=PLOT_BROKEN_IDS)
+    @pytest.mark.parametrize("xdata", fvalues, ids=fids)
+    def test_plot_failures(self, ax, plotter, xdata):
+        with pytest.raises(TypeError):
```

Comment:
Would it be better to mark these as expected fail instead of catching the error and passing? That way when the test suite is run we know it's something that should really work, but doesn't yet.

Reply:
The failure is that they don't throw the TypeError, and those plots are marked xfail (and report as such when running the tests).

Reply:
Ah sorry, didn't see the xfail above, all is good :)